src/queries/test-id.js

```diff
@@ -11,8 +12,24 @@ function queryAllByTestId(...args) {
 
 const getMultipleError = (c, id) =>
   `Found multiple elements by: [${getTestIdAttribute()}="${id}"]`
-const getMissingError = (c, id) =>
-  `Unable to find an element by: [${getTestIdAttribute()}="${id}"]`
+const getMissingError = (
+  c,
+  id,
+  {computeCloseMatches = false, ...options} = {},
+) => {
+  const defaultMessage = `Unable to find an element by: [${getTestIdAttribute()}="${id}"]`
+
+  const closeMatches =
+    !computeCloseMatches || typeof id !== 'string'
+      ? []
+      : getCloseMatchesByAttribute(getTestIdAttribute(), c, id, options)
```

Comment:
I'm concerned about this increasing the performance issues we already have with `find*` queries which are expected to fail at least once. Any chance we could lazily calculate this value so it's only run when the error is actually displayed? I don't know whether this is possible.

But perhaps my concern is unwarranted? Maybe this is faster than I think?

Reply:
> I'm concerned about this increasing the performance issues we already have with find* queries which are expected to fail at least once.

Good point. I tested a few times on my local running `getByTestId('search', {computeCloseMatches: true})` vs `getByTestId('search', {computeCloseMatches: false})`. 

When false, it finished consistently at around 20ms. When true finished at around 35ms. It increases with the number of elements found, but not by much. Might become slightly quicker with the [recommended lib](https://github.com/sindresorhus/leven)

> Any chance we could lazily calculate this value so it's only run when the error is actually displayed? 

An alternative would be to throw functions for `find*` queries. And then check if `typeof lastError === 'function'`  when [waitFor](https://github.com/testing-library/dom-testing-library/blob/master/src/wait-for.js#L156) times out. 

That might mean decoupling `find*` and `get*` queries a bit or perhaps make `get` queries depend on `find` queries instead of the other way around. 

i.e: a `get*` query could be a `find*` query with `{timeout: 0, interval: 0}` ? (not sure if that would work)

Since that seems a bit involved, we could start with a config `computeCloseMatches` that defaults to `false` and give that some testing?

Reply:
> I'm concerned about this increasing the performance issues we already have with find* queries which are expected to fail at least once.

Good point. I tested a few times on my local running `getByTestId('search', {computeCloseMatches: true})` vs `getByTestId('search', {computeCloseMatches: false})`. 

Not a reliable benchmark, but when false it finished consistently at around 20ms. When true finished at around 35ms. It increases with the number of elements found, but not by much. Might become slightly quicker with the [recommended lib](https://github.com/sindresorhus/leven). 

> Any chance we could lazily calculate this value so it's only run when the error is actually displayed?

An alternative would be to throw functions for `find*` queries. And then check if `typeof lastError === 'function'`  when [waitFor](https://github.com/testing-library/dom-testing-library/blob/master/src/wait-for.js#L156) times out.

That might mean decoupling `find*` and `get*` queries a bit or perhaps make `get` queries depend on `find` queries instead of the other way around.

i.e: a `get*` query could be a `find*` query with `{timeout: 0, interval: 0}` ? (not sure if that would work)

Since that seems a bit involved, we could start with a config `computeCloseMatches` that defaults to `false` and give that some testing?

Reply:
I'm ok with giving it a trial run and seeing what real-world experience with it will be like, so defaulting to disabled makes sense to me. Let's try it out using `leven`.

Reply:
@kentcdodds do you think we should start with the testId query or add this feature to all 10 queries? Wondering if i can break down the work somehow...

Re. the performance concern. Maybe if we realise there is a huge performance impact we can skip the computation of close matches on `find*` queries(similar to what was done for the `role` query here: https://github.com/testing-library/dom-testing-library/pull/590

Reply:
I'm not sure. What does everyone else think?

Reply:
IMO if the default is false, we can give it a try in more queries, though, this seems to me like a configuration that shouldn't be set to true by default at all, especially since it has a performance impact. I see this as something that will not be helpful in CI for example and will only take more time so if the developer wants to opt in they will have an option to do that.
Putting aside what I wrote above, I really like this PR and do think it can have a valuable impact so thanks for this :)

Reply:
Can you please try a benchmark again using the new `leven` implementation?